Domain/Transaction.cs

```diff
@@ -0,0 +1,196 @@
+﻿using System;
+using System.Collections.Generic;
+
+using Types;
+
+namespace Domain
+{
+    public abstract class Transaction       
+    {
+        private Transaction(CounterParty counterParty, BankAccount account, decimal amount, DateTime receivedDateTime, string description, int sequence, ICollection<TransactionAttachment> attachments)        {
+            Id = Guid.NewGuid();
+            DateFiled = DateTime.UtcNow;
+
+            CounterParty = counterParty ?? throw new ArgumentNullException(nameof(counterParty), "Cannot be null");
+            Account = account ?? throw new ArgumentNullException(nameof(account), "Cannot be null");
```

Comment:
I don''t think `null` checks will suffice here, we want to make sure we have a meaningful value, so an empty string should also throw an exception